File: SalmonTrollingAssistant/Domain/Services/BatteryOptimizationManager.swift
import Foundation
import UIKit
import Combine

/**
 * Manages battery optimization strategies across the app
 */
class BatteryOptimizationManager {
    // Singleton instance
    static let shared = BatteryOptimizationManager()
    
    // Battery optimization modes
    enum OptimizationMode: String {
        case performance // Prioritize app performance over battery life
        case balanced    // Balance between performance and battery life
        case batterySaver // Maximize battery life, potentially sacrificing some features
    }
    
    // UserDefaults keys
    private let optimizationModeKey = "battery_optimization_mode"
    private let locationUpdateIntervalKey = "location_update_interval"
    private let networkBatchWindowKey = "network_batch_window"
    private let lowBatteryThresholdKey = "low_battery_threshold"
    private let lastBatteryLevelKey = "last_battery_level"
    private let lastBatteryCheckKey = "last_battery_check"
    
    // Publishers
    private let optimizationModeSubject = CurrentValueSubject<OptimizationMode, Never>(.balanced)
    private let batteryLevelSubject = CurrentValueSubject<Int, Never>(100)
    private let isChargingSubject = CurrentValueSubject<Bool, Never>(false)
    
    // Public publishers
    var optimizationMode: AnyPublisher<OptimizationMode, Never> {
        return optimizationModeSubject.eraseToAnyPublisher()
    }
    
    var batteryLevel: AnyPublisher<Int, Never> {
        return batteryLevelSubject.eraseToAnyPublisher()
    }
    
    var isCharging: AnyPublisher<Bool, Never> {
        return isChargingSubject.eraseToAnyPublisher()
    }
    
    private init() {
        // Load saved mode from UserDefaults
        if let savedModeString = UserDefaults.standard.string(forKey: optimizationModeKey),
           let savedMode = OptimizationMode(rawValue: savedModeString) {
            optimizationModeSubject.send(savedMode)
        }
        
        // Start monitoring battery
        setupBatteryMonitoring()
        
        // Initial battery check
        updateBatteryStats()
    }
    
    // MARK: - Public Methods
    
    /**
     * Set the battery optimization mode
     */
    func setOptimizationMode(_ mode: OptimizationMode) {
        optimizationModeSubject.send(mode)
        UserDefaults.standard.set(mode.rawValue, forKey: optimizationModeKey)
    }
    
    /**
     * Get the location update interval based on current optimization mode and battery level
     * @return Update interval in seconds
     */
    func getLocationUpdateInterval() -> TimeInterval {
        let mode = optimizationModeSubject.value
        let level = batteryLevelSubject.value
        
        // Base intervals for different modes
        switch (mode, level) {
        case (_, 0...15):
            return 60.0 // 1 minute when battery is low, regardless of mode
        case (.performance, _):
            return 15.0 // 15 seconds in performance mode
        case (.balanced, _):
            return 30.0 // 30 seconds in balanced mode
        case (.batterySaver, _):
            return 60.0 // 1 minute in battery saver mode
        }
    }
    
    /**
     * Get the network batch window - time to wait to batch network requests
     * @return Batch window in seconds
     */
    func getNetworkBatchWindow() -> TimeInterval {
        let mode = optimizationModeSubject.value
        
        switch mode {
        case .performance:
            return 1.0 // 1 second in performance mode
        case .balanced:
            return 5.0 // 5 seconds in balanced mode
        case .batterySaver:
            return 15.0 // 15 seconds in battery saver mode
        }
    }
    
    /**
     * Get the current low battery threshold
     */
    func getLowBatteryThreshold() -> Int {
        return UserDefaults.standard.integer(forKey: lowBatteryThresholdKey)
    }
    
    /**
     * Set the low battery threshold
     */
    func setLowBatteryThreshold(_ threshold: Int) {
        let validThreshold = min(max(threshold, 5), 50)
        UserDefaults.standard.set(validThreshold, forKey: lowBatteryThresholdKey)
    }
    
    /**
     * Automatically adjust optimization mode based on device state
     */
    func autoAdjustOptimizationMode() {
        let level = batteryLevelSubject.value
        let charging = isChargingSubject.value
        let lowPowerMode = ProcessInfo.processInfo.isLowPowerModeEnabled
        
        switch (charging, lowPowerMode, level) {
        case (true, _, _):
            setOptimizationMode(.performance)
        case (_, true, _):
            setOptimizationMode(.batterySaver)
        case (_, _, 0...20):
            setOptimizationMode(.batterySaver)
        case (_, _, 21...49):
            setOptimizationMode(.balanced)
        case (_, _, 50...100):
            setOptimizationMode(.balanced)
        default:
            setOptimizationMode(.balanced)
        }
    }
    
    // MARK: - Private Methods
    
    private func setupBatteryMonitoring() {
        // Enable battery monitoring
        UIDevice.current.isBatteryMonitoringEnabled = true
        
        // Observe battery level changes
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(batteryLevelDidChange),
            name: UIDevice.batteryLevelDidChangeNotification,
            object: nil
        )
        
        // Observe battery state changes (charging/discharging)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(batteryStateDidChange),
            name: UIDevice.batteryStateDidChangeNotification,
            object: nil
        )
        
        // Observe low power mode changes
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(lowPowerModeDidChange),
            name: .NSProcessInfoPowerStateDidChange,
            object: nil
        )
    }
    
    @objc private func batteryLevelDidChange() {
        updateBatteryStats()
        autoAdjustOptimizationMode()
    }
    
    @objc private func batteryStateDidChange() {
        let isCurrentlyCharging = UIDevice.current.batteryState == .charging || 
                                 UIDevice.current.batteryState == .full
        isChargingSubject.send(isCurrentlyCharging)
        autoAdjustOptimizationMode()
    }
    
    @objc private func lowPowerModeDidChange() {
        autoAdjustOptimizationMode()
    }
    
    private func updateBatteryStats() {
        let currentLevel = Int(UIDevice.current.batteryLevel * 100)
        batteryLevelSubject.send(currentLevel)
        
        UserDefaults.standard.set(currentLevel, forKey: lastBatteryLevelKey)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: lastBatteryCheckKey)
    }
}

File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/data/service/BatteryOptimizationManager.kt
package com.example.salmontrollingassistant.data.service

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.PowerManager
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private val Context.batteryOptimizationStore by preferencesDataStore(name = "battery_optimization_settings")

/**
 * Manages battery optimization strategies across the app
 */
@Singleton
class BatteryOptimizationManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    // Keys for battery optimization settings
    private val BATTERY_OPTIMIZATION_MODE = stringPreferencesKey("battery_optimization_mode")
    private val LOCATION_UPDATE_INTERVAL = longPreferencesKey("location_update_interval")
    private val NETWORK_BATCH_WINDOW = longPreferencesKey("network_batch_window")
    private val LOW_BATTERY_THRESHOLD = intPreferencesKey("low_battery_threshold")
    private val LAST_BATTERY_LEVEL = intPreferencesKey("last_battery_level")
    private val LAST_BATTERY_CHECK = longPreferencesKey("last_battery_check")

    // Battery optimization modes
    enum class OptimizationMode {
        PERFORMANCE, // Prioritize app performance over battery life
        BALANCED,    // Balance between performance and battery life
        BATTERY_SAVER // Maximize battery life, potentially sacrificing some features
    }

    /**
     * Get the current battery optimization mode
     */
    fun getOptimizationMode(): Flow<OptimizationMode> {
        return context.batteryOptimizationStore.data.map { preferences ->
            val modeString = preferences[BATTERY_OPTIMIZATION_MODE] ?: OptimizationMode.BALANCED.name
            try {
                OptimizationMode.valueOf(modeString)
            } catch (e: Exception) {
                OptimizationMode.BALANCED
            }
        }
    }

    /**
     * Set the battery optimization mode
     */
    suspend fun setOptimizationMode(mode: OptimizationMode) {
        context.batteryOptimizationStore.edit { preferences ->
            preferences[BATTERY_OPTIMIZATION_MODE] = mode.name
        }
    }

    /**
     * Get the location update interval based on current optimization mode and battery level
     * @return Update interval in milliseconds
     */
    suspend fun getLocationUpdateInterval(): Long {
        val mode = getOptimizationMode().map { it }.collect { it }
        val batteryLevel = getBatteryLevel()
        
        // Base intervals for different modes
        return when {
            batteryLevel <= 15 -> 60000L // 1 minute when battery is low, regardless of mode
            mode == OptimizationMode.PERFORMANCE -> 15000L // 15 seconds in performance mode
            mode == OptimizationMode.BALANCED -> 30000L // 30 seconds in balanced mode
            mode == OptimizationMode.BATTERY_SAVER -> 60000L // 1 minute in battery saver mode
            else -> 30000L // Default to 30 seconds
        }
    }

    /**
     * Get the network batch window - time to wait to batch network requests
     * @return Batch window in milliseconds
     */
    suspend fun getNetworkBatchWindow(): Long {
        val mode = getOptimizationMode().map { it }.collect { it }
        
        return when (mode) {
            OptimizationMode.PERFORMANCE -> 1000L // 1 second in performance mode
            OptimizationMode.BALANCED -> 5000L // 5 seconds in balanced mode
            OptimizationMode.BATTERY_SAVER -> 15000L // 15 seconds in battery saver mode
        }
    }

    /**
     * Get the current battery level
     * @return Battery level percentage (0-100)
     */
    fun getBatteryLevel(): Int {
        val batteryIntent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val level = batteryIntent?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
        val scale = batteryIntent?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1
        
        return if (level != -1 && scale != -1) {
            (level * 100 / scale.toFloat()).toInt()
        } else {
            -1
        }
    }

    /**
     * Check if the device is in power save mode
     */
    fun isInPowerSaveMode(): Boolean {
        val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        return powerManager.isPowerSaveMode
    }

    /**
     * Check if the device is charging
     */
    fun isCharging(): Boolean {
        val batteryIntent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        val status = batteryIntent?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
        
        return status == BatteryManager.BATTERY_STATUS_CHARGING || 
               status == BatteryManager.BATTERY_STATUS_FULL
    }

    /**
     * Automatically adjust optimization mode based on device state
     */
    suspend fun autoAdjustOptimizationMode() {
        when {
            isCharging() -> setOptimizationMode(OptimizationMode.PERFORMANCE)
            isInPowerSaveMode() -> setOptimizationMode(OptimizationMode.BATTERY_SAVER)
            getBatteryLevel() <= 20 -> setOptimizationMode(OptimizationMode.BATTERY_SAVER)
            getBatteryLevel() >= 50 -> setOptimizationMode(OptimizationMode.BALANCED)
        }
    }

    /**
     * Update the stored battery level and check time
     */
    suspend fun updateBatteryStats() {
        val currentLevel = getBatteryLevel()
        val currentTime = System.currentTimeMillis()
        
        context.batteryOptimizationStore.edit { preferences ->
            preferences[LAST_BATTERY_LEVEL] = currentLevel
            preferences[LAST_BATTERY_CHECK] = currentTime
        }
    }

    /**
     * Get the low battery threshold percentage
     */
    fun getLowBatteryThreshold(): Flow<Int> {
        return context.batteryOptimizationStore.data.map { preferences ->
            preferences[LOW_BATTERY_THRESHOLD] ?: 15
        }
    }

    /**
     * Set the low battery threshold percentage
     */
    suspend fun setLowBatteryThreshold(threshold: Int) {
        context.batteryOptimizationStore.edit { preferences ->
            preferences[LOW_BATTERY_THRESHOLD] = threshold.coerceIn(5, 50)
        }
    }
}